src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs

```diff
@@ -1094,53 +1096,71 @@ internal static StackValue[] GenerateRangeByAmount(decimal start, long amount, d
                 return StackValue.Null;
             }
 
-            double startValue = svStart.ToDouble().RawDoubleValue;
-            if (double.IsInfinity(startValue) || double.IsNaN(startValue))
+            StackValue[] range = null;
+
+            if (svStart.IsNumeric)
             {
-                runtimeCore.RuntimeStatus.LogWarning(
-                    WarningID.kInvalidArguments, 
-                    Resources.kInvalidArgumentsInRangeExpression);
-                return StackValue.Null;
+                range = GenerateNumericRange(
+                    svStart,
+                    svEnd,
+                    svStep,
+                    svOp,
+                    hasStep,
+                    hasAmountOp,
+                    runtimeCore);
+            }
+            else
+            {
+                if (svStart.IsString)
+                {
+                    range = GenerateAlphabetRange(
+                    svStart,
+                    svEnd,
+                    svStep,                   
+                    runtimeCore);
+                }
             }
-            decimal start = new decimal(startValue);
 
+            return range == null ? StackValue.Null : runtimeCore.RuntimeMemory.Heap.AllocateArray(range);
+        }
+
+        private static StackValue[] GenerateNumericRange(StackValue svStart,
```

Comment:
`GenerateNumericRange` is almost the same as it was. I decided to separate current code in 2 different methods for numerical range and alphabet range, because I wanted to make code more clean.